File: A2Z_striver/patterns/pattern_2/pattern_2.py
import sys

class Solution:
    """
    Problem: Print a Right-Angled Triangle of stars.
    Example Input: n = 4
    Example Output:
    *
    **
    ***
    ****
    
    Constraints: 1 <= n <= 100
    Goal: O(N^2) Time.
    """

    # // ~ Approach 1: Pythonic (String Multiplication)
    # // Analysis: Time O(N^2) | Space O(1)
    def print_increasing_triangle_pythonic(self, n):
        for i in range(n):
            # // . Logic: Row 0 needs 1 star, Row 1 needs 2 stars...
            print("*" * (i + 1))

    # // ~ Approach 2: Nested Loops (Universal Logic)
    # // Analysis: Time O(N^2) | Space O(1)
    def print_increasing_triangle_nested(self, n):
        # // ! This logic translates directly to C++/Java
        for i in range(n):
            # // - Inner loop runs (i + 1) times
            for j in range(i + 1):
                print("*", end="")
            print()

# Driver Code
if __name__ == "__main__":
    solver = Solution()
    N = 5
    
    print(f"--- Approach 1: Pythonic (N={N}) ---")
    solver.print_increasing_triangle_pythonic(N)
    
    print(f"\n--- Approach 2: Nested Loops (N={N}) ---")
    solver.print_increasing_triangle_nested(N)

File: A2Z_striver/1 Basics/1.2 Patterns/pattern_2/pattern_2.cpp
#include <iostream>

using namespace std;

// Header Comment Block
// Problem: Print a Right-Angled Triangle of stars.
// Example Input: n = 4
// Example Output:
// *
// **
// ***
// ****
// Constraints: 1 <= n <= 100
// Goal: O(N^2) Time | O(1) Space

class Solution
{
public:
    // ~ Approach: Dependent Nested Loops
    // Analysis:
    // Time: O(N^2) - Specifically N(N+1)/2 operations.
    // Space: O(1).
    void print_increasing_triangle(int n)
    {
        // - Outer Loop: Rows
        for (int i = 0; i < n; i++)
        {

            // - Inner Loop: Columns
            // ! Key Logic: j runs from 0 to i.
            // - When i=0, loop runs 1 time.
            // - When i=1, loop runs 2 times.
            for (int j = 0; j <= i; j++)
            {
                cout << "*";
            }

            // ! Performance: Use '\n'
            cout << "\n";
        }
    }
};

int main()
{
    // ! Fast I/O
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;
    int n = 5;

    cout << "Test Case (n=" << n << "):\n";
    sol.print_increasing_triangle(n);

    return 0;
}